working/augmentation-libraries/feature-specification.md

```diff
@@ -697,14 +697,21 @@ It is a compile-time error if:
 
 ### Augmenting constructors
 
-Constructors are (as always) more complex. A constructor marked `augment`
-replaces the body of the existing constructor with its body. If the augmenting
-constructor has any initializers, they are appended to the original
-constructor's initializers, but before any original super initializer or
-original redirecting initializer if there is one.
-
-In the augmenting constructor's body, an `augmented()` call invokes the
-original constructor's body.
+Constructors are (as always) more complex. A constructor marked `augment` may:
+
+- Add or replace the body of the existing constructor with a new body.
+  - If the augmenting constructor has no explicit body (terminates with a `;`),
+    then the original is left in place.
+  - In the augmenting constructor's body, an `augmented()` call invokes the
+    original constructor's body. Unlike regular function augmentations, you do
+    not pass parameters explicitly. All of the same parameters are in scope in
+    the augmented body as the augmenting body.
+    - If a parameter variable is overwritten prior to calling `augmented()`, the
+      augmented body will see the updated value. All references to the parameter
+      in the augmented and augmenting bodies refer to the exact same variable.
+- Add initializers to the initializer list.
+  - These are appended to the original constructor's initializers, but before
+    any super initializer or redirecting initializer (if present).
```

Comment:
As written, it says you can add initializer list entries before a redirecting initializer. That's not valid. A redirecting (generative) constructor cannot have initializer list entries at all. (It can't even have `asserts` according to spec, but it seems implementations allow that, which is kind of nice.)

I'll assume (or propose if not true) that an augmenting constructor must have the same *kind* (out of the seven possible) as the constructor it augments. (Unless it can add `const`, which may be possible in some cases, but definitely not all.)

* Augmenting constructor must be `const` if and only if augmented constructor is `const`.
* Augmetning constructor must be `factory` if and only if augmented constructor is `factory`.
* Augmenting constructor must be redirecting if and only if augmented constructor is redirecting.

Then:
* A non-redirecting factory constructor can augment the body of the augmented constructor. This works just like a function body augementation, and you can invoke the augmented body as `augmented(arguments)`.
* A redirecting factory constructor can augment the redirection target constructor of the augmented constructor. This replaces the target with the augmenting constructor's target. (That target must redirect be a constructor that creates a subtype of the target of the augmented constructor?)
* A redirecting generative constructor can augment the target constructor invocation of the augmented constructor. This replaces the target with the target constructor invocation of the augmented constructor. It cannot use `augmented`.
    * Maybe it can also add asserts to the initializer list. These are executed after any asserts
       of the augmented constructor.
* A non-redirecting generative constuctor can augment the parameter list, initializer list, 
   super constructor invocation and body of the augmented constructor.
    * Changing a parameter to be an initializing formal removes existing initializer list entries
       for that variable.
    * Changing a parameter to not be an initializing formal then allows (possibly requires) it 
       to be initialized in the initializer list.
    * Changing a parameter to be a `super.` parameter replaces a corresponding argument
       from the existing super-constructor invocation, if there is one.
    * Changing a parameter away from being `super.` makes the super-constructor invocation
       not have a parameter at that position. This likely makes the invocation invalid, and 
       having to be replaced.
    * Changing the name of a positional parameter is allowed.
       Whenever an expression or body of the augmented constructor is evaluated, it happens 
       in a *corresponding scope*, which is a scope which binds the same variables, but with
       the names that they would have had in the augmented constructor.
    * If the augmenting constructor has an initializer list entry for the instance variable *n*, then:
        * If the augmented constructor (the result of all prior augmentations) has no initializer
           for *n*, then the initializer is appended (evaluated after all existing initializer lists).
        * If the augmented constructor already has an initializer for *n*, 
           then the augmenting initializer replaces that initializer's initializer expression,
           and `augmented` evaluates the replaced initializer expression in the corresponding scope.
    * If the augmenting constructor has a body, it replaces the body of the augmented 
       constructor.
       It may use `augmented` as an expression with static type `void`, which executes
       the replaced body in its corresponding scope.

In all of these, if the constructors are `const`, any inserted expressions must be valid for that. (Which just means that augmenting an initializer list entry can safely refer to its `augmented`, because that's known to be potentially constant.)

Reply:
I totally re-wrote this section, PTAL